src/Dialect/ONNX/IndexExpr.cpp

```diff
@@ -0,0 +1,1062 @@
+//===----------------IndexExpr.cpp - Index expression---------------------=== //
+//
+// copyright 2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file handle index expressions using indices and calcualtions using
+// literals, affine expressions, and values.
+//
+//===----------------------------------------------------------------------===//
+
+// both debug variables will be removed once debugging is complete.
+#define DEBUG 0
+#define CEIL_FLOOR_IN_STD 0
+
+#include "src/Dialect/ONNX/IndexExpr.hpp"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/MathExtras.h"
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/Sequence.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// IndexExprContext constructors.
+//===----------------------------------------------------------------------===//
+
+IndexExprContext::IndexExprContext(
+    ConversionPatternRewriter *rewriter, Location loc)
+    : rewriter(rewriter), loc(loc), dims(), symbols(), parentContext(nullptr) {}
+
+IndexExprContext::IndexExprContext(IndexExprContext &newParentContext)
+    : rewriter(newParentContext.rewriter), loc(newParentContext.loc), dims(),
+      symbols(), parentContext(nullptr) {
+  // We resue the parent context, and in particuliar its affine
+  // functions. Now because the affine functions of the parent context have
+  // "ids" embedded in the AffineExpr, we must reuse the same mix of Dims and
+  // Symbols here. I don't believe there is any sideeffects in considering a Dim
+  // from the parent's context as a Dim in the child's context, even though the
+  // parent's dim is supposed to be constant in the child's context.
+  for (Value parentDim : newParentContext.dims)
+    addDim(parentDim);
+  for (Value parentSymbol : newParentContext.symbols)
+    addSymbol(parentSymbol);
+  // Save reference to parent context so that we may detect the reuse.
+  parentContext = &newParentContext;
+}
+
+//===----------------------------------------------------------------------===//
+// IndexExprContext builder for IndexExpr.
+//===----------------------------------------------------------------------===//
+
+IndexExpr IndexExprContext::createUndefinedIndex() {
+  IndexExpr res;
+  return res.initAsUndefined();
```

Comment:
It seems to me that the res.initasUndefined() returns the object res. Will this cause any memory issue?